File: java-aiproxy/src/main/java/com/aiproxy/provider/OpenAIProvider.java
package com.aiproxy.provider;

import com.aiproxy.model.Protocol;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

@Slf4j
@Component
public class OpenAIProvider implements AIProvider {
    
    private final WebClient webClient;
    
    public OpenAIProvider(
            @Value("${openai.api-key:}") String apiKey,
            @Value("${openai.base-url:https://api.openai.com/v1}") String baseUrl,
            WebClient.Builder webClientBuilder) {
        this.webClient = webClientBuilder
                .baseUrl(baseUrl)
                .defaultHeader("Authorization", "Bearer " + apiKey)
                .defaultHeader("Content-Type", "application/json")
                .build();
    }
    
    @Override
    public Mono<Object> chatCompletion(Object request) {
        return webClient.post()
                .uri("/chat/completions")
                .contentType(MediaType.APPLICATION_JSON)
                .bodyValue(request)
                .retrieve()
                .bodyToMono(Object.class)
                .doOnError(e -> log.error("OpenAI API error", e));
    }
    
    @Override
    public Protocol getProtocol() {
        return Protocol.OPENAI;
    }
    
    @Override
    public String getName() {
        return "openai";
    }
}

File: scala-aiproxy/src/main/scala/com/aiproxy/provider/OpenAIProvider.scala
package com.aiproxy.provider

import com.aiproxy.model.{Protocol, OpenAI}
import sttp.client3.*
import sttp.client3.akkahttp.AkkaHttpBackend
import spray.json.*
import scala.concurrent.{Future, ExecutionContext}
import com.typesafe.scalalogging.LazyLogging

class OpenAIProvider(
  apiKey: String,
  baseUrl: String = "https://api.openai.com/v1"
)(using ec: ExecutionContext) extends AIProvider with LazyLogging:

  private val backend = AkkaHttpBackend()

  def chatCompletion(request: Any): Future[Any] =
    import OpenAI.given
    
    request match
      case req: OpenAI.Request =>
        val requestBody = req.toJson.compactPrint
        
        val response = basicRequest
          .post(uri"$baseUrl/chat/completions")
          .header("Authorization", s"Bearer $apiKey")
          .header("Content-Type", "application/json")
          .body(requestBody)
          .send(backend)

        response.map { resp =>
          logger.info(s"OpenAI response status: ${resp.code}")
          resp.body match
            case Right(body) => body.parseJson.convertTo[OpenAI.Response]
            case Left(error) => throw new Exception(s"OpenAI API error: $error")
        }
      
      case _ => Future.failed(new Exception("Invalid request type"))

  def protocol: Protocol = Protocol.OpenAI
  def name: String = "openai"
